### PR TITLE
Fix: remove CURLRequest headers sharing from $_SERVER

### LINKED ISSUE #4826: Bug: Extra Headers on CURLRequest
**Describe the bug**
When you make a POST call to any controller on ServerA (RequestA) and then you use \Config\Services::curlrequest to make a "GET" request to a ServerB (RequestB), "MessageTrait.php" attach all the  RequestA Headers from the POST to the RequestB.  So the header "Content-Length" is sent to the ServerB.

**CodeIgniter 4 version**
CI 4.1.3

**Affected module(s)**
system/HTTP/CURLRequest.php
system/HTTP/MessageTrait.php

**Expected behavior, and steps to reproduce if appropriate**
The header "Content-Length" from the RequestA is sent to the ServerB as a Header.
Then ServerB wait for the body of the RequestB but since RequesB is "GET" the body is <EMPTY> that means the server make a GatewayTimeout.

**Context**
 - PHP version 7.4.20

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
Alternative to #5218
Fixes #4826
- `populateHeaders()` which gets request headers from the browser and sets the same headers to the curl request and shares the headers between curl requests, is a bug. So remove `populateHeaders()` logic completely
- add `Config\CURLRequest::$shareOptions`. The default is `true`
  - if `false`,
    - all options are reset after sending a request
    - `$defaultOptions` which are passed in the constructor are applied to all requests
- ~~share `$option['headers']` except `$unsharedHeaders` between curl requests~~
- remove `CURLRequestTest::testOptionsHeadersUsingPopulate()`

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Review by michalsn — COMMENTED** [review]
Nice work @kenjis, I think it looks really good.

The only important thing that is missing is the separate config file and some documentation for this new feature (`shareOptions`).

**Comment by michalsn** [comment]
> * let's use other namespace to keep BC
> * at first need create class skeletons (architecture)
>   2.1. extract curl options in "config" class

@WinterSilence I'm not sure if I'm understanding you correctly, but if you're proposing to add a new implementation of this class because of BC, that's not going to happen. The possible BC break is accidental and is a derivative from a bug that is now fixed.

Technically speaking we are dealing with a bug here. However, we are aware that fixing it may lead to problems in some cases. Therefore, we raise this issue in the changelog and upgrade notes.

File: app/Config/App.php [hunk 1]
```diff
@@ -461,4 +461,15 @@ class App extends BaseConfig
...
+    /**
+     * --------------------------------------------------------------------------
+     * CURLRequest Share Options
+     * --------------------------------------------------------------------------
+     *
+     * Whether share options between requests or not.
+     *
+     * @var bool
+     */
+    public $CURLRequestShareOptions = true;
```
[the remark this answers is not in the document]
Comment (WinterSilence) [reply app/Config/App.php:474]: @michalsn

> only one variable now

single in App, but we must declare CURLRequest properties too

File: system/HTTP/CURLRequest.php [hunk 2]
```diff
@@ -72,6 +79,20 @@ class CURLRequest extends Request
...
+     * @var array
+     */
+    private $defaultOptions;
+
+    /**
+     * Whether share options between requests or not.
+     *
+     * @var bool
+     */
+    private $shareOptions;
```
[the remark this answers is not in the document]
Comment (kenjis) [reply system/HTTP/CURLRequest.php:94]: `$defaultOptions` is set from the constructor param.
`shareOptions` is set form the config file.

I think they don't need to be overridden.

File: app/Config/App.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (kenjis) [reply app/Config/App.php]: Moved to `Config\CURLRequest`.

File: system/HTTP/CURLRequest.php [hunk 3]
```diff
@@ -42,7 +42,14 @@ class CURLRequest extends Request
...
      */
-    protected $config = [
+    protected $config;
+
+    /**
+     * The default setting values
+     *
+     * @var array
+     */
+    private $defaultConfig = [
```
[the remark this answers is not in the document]
Comment (kenjis) [reply system/HTTP/CURLRequest.php]: Fiixed.

File: system/HTTP/CURLRequest.php [hunk 4]
```diff
@@ -114,9 +138,26 @@ public function request($method, string $url, array $options = []): ResponseInte
...
         $this->send($method, $url);
 
+        if ($this->shareOptions === false) {
+            $this->resetOptions();
+        }
+
         return $this->response;
     }
 
+    private function resetOptions()
```
[the remark this answers is not in the document]
Comment (kenjis) [reply system/HTTP/CURLRequest.php]: Fixed.

**Comment by kenjis** [comment]
@WinterSilence I'm also not get what you want to say.

**Comment by WinterSilence** [comment]
@kenjis I want minimal changes in "behavior" of current classes i.e. save public part "as is". Next release is not new major version - BC not allowed. But some errors can't be solved without BC, so I suggest a temporary solution - similar classes with fixes - users will choose version to use. Old/deprecated classes will be removed in next major version.

File: system/HTTP/CURLRequest.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (michalsn) [reply system/HTTP/CURLRequest.php]: I still think that users can come up with some scenarios when this change would be beneficial but... fair enough. I might be overthinking this.

**Review by michalsn — APPROVED** [review]
From my point of view, this is ready. Thanks!

**Review by MGatner — APPROVED** [review]

**Comment by kenjis** [comment]
@WinterSilence I thought that old/new two classes.
If we have two classes, we have to maintain two classes.

Adding config for a new behavior is a bit better than maintaining two classes, I thought.
And users can use the new behavior, just set the config.

**Pull request merged**